improver_tests/expected_value/test_expected_value.py

```diff
@@ -0,0 +1,113 @@
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the ExpectedValue plugin."""
+import numpy as np
+import pytest
+from iris.coords import CellMethod
+from iris.exceptions import CoordinateNotFoundError
+from numpy.testing import assert_allclose
+
+from improver.expected_value import ExpectedValue
+from improver.synthetic_data.set_up_test_cubes import (
+    set_up_percentile_cube,
+    set_up_probability_cube,
+    set_up_variable_cube,
+)
+
+
+@pytest.fixture
+def realizations_cube():
+    data = np.array(
+        [range(0, 9), range(10, 19), range(30, 39)], dtype=np.float32
+    ).reshape([3, 3, 3])
+    return set_up_variable_cube(data, realizations=[0, 1, 2])
+
+
+@pytest.fixture
+def percentile_cube():
+    data = np.array(
+        [range(10, 19), range(20, 29), range(40, 49)], dtype=np.float32
+    ).reshape([3, 3, 3])
+    return set_up_percentile_cube(data, percentiles=[25, 50, 75])
+
+
+@pytest.fixture
+def threshold_cube():
+    probs = np.array([1.0, 0.7, 0.5, 0.45, 0.0], dtype=np.float32)
+    data = np.broadcast_to(probs[:, np.newaxis, np.newaxis], [5, 3, 2])
+    return set_up_probability_cube(data, thresholds=[280, 281, 282, 283, 284])
+
+
+def test_process_realizations_basic(realizations_cube):
+    """Check that the expected value of realisations calculates the mean and
+    appropriately updates metadata."""
+    expval = ExpectedValue().process(realizations_cube)
+    # coords should be the same, but with the first (realization) dimcoord removed
+    assert expval.coords() == realizations_cube.coords()[1:]
+    # a cell method indicating mean over realizations should be added
+    assert expval.cell_methods == (CellMethod("mean", "realization"),)
+    # mean of the realisation coord (was the first dim of data)
+    expected_data = np.mean(realizations_cube.data, axis=0)
+    assert_allclose(expval.data, expected_data)
+
+
+def test_process_percentile_basic(percentile_cube):
+    """Check that percentiles are converted to realisations and the mean is
+    calculated."""
+    expval = ExpectedValue().process(percentile_cube)
+    # coords should be the same, but with the first (percentile) dimcoord removed
+    assert expval.coords() == percentile_cube.coords()[1:]
+    # a cell method indicating mean over realizations should be added
+    assert expval.cell_methods == (CellMethod("mean", "realization"),)
+    # this works out to be a mean over percentiles
+    # since the percentiles are equally spaced
```

Comment:
That's a good point. There is an assumption in the rebadging code that the percentiles are equally spaced and centred on 50%, but I can't see a check for that. It would be worthwhile raising an issue to explicitly check input data to RebadgePercentilesAsRealizations, or at least document the expectation.

Reply:
I was going to raise this issue too, before I noticed this thread. I think we've always got away with it before because we've always used it after resampling percentile (e.g. https://github.com/metoppv/improver/blob/master/improver/cli/generate_realizations.py#L95), but I think you're right that the rebadge plugin should fail if the input percentiles aren't evenly spaced quantiles. Do you want to consider putting in the resample percentiles step in this PR? Or just do something in the new issue #1724?

Reply:
I think it's best to tackle this topic as part of #1724 in order to avoid scope creep in this PR.